text/0000-cargo-prepublish.md

```diff
@@ -0,0 +1,395 @@
+entries in the `[augment]` section to precise dependencies, encoding them all in
+the lock file whether they're used or not.
+
+Dependencies formed on crates listed in `[augment]` will then be listed directly
+in Cargo.lock, eschewing the actual dependency entirely. For example let's say
+we depend on `env_logger` but we're using `[augment]` to depend on a git version
+of the `log` crate, a dependency of `env_logger`. First we'll have our
+`Cargo.toml` including:
+
+```toml
+# Cargo.toml
+[dependencies]
+env_logger = "0.4"
+```
+
+With that we'll find this in `Cargo.lock`, notably everything comes from
+crates.io
+
+```toml
+# Cargo.lock
+[[package]]
+name = "env_logger"
+version = "0.4.2"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+dependencies = [
+ "log 0.3.7 (registry+https://github.com/rust-lang/crates.io-index)",
+]
+
+[[package]]
+name = "log"
+version = "0.3.7"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+```
+
+Next up we'll add our `[augment]` section to crates.io:
+
+```toml
+# Cargo.toml
+[augment.crates-io]
+log = { git = 'https://github.com/rust-lang-nursery/log' }
+```
+
+and that will generate a lock file that looks (roughly) like:
+
+```toml
+# Cargo.lock
+[[package]]
+name = "env_logger"
+version = "0.4.2"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+dependencies = [
+ "log 0.3.7 (git+https://github.com/rust-lang-nursery/log)",
+]
+
+[[package]]
+name = "log"
+version = "0.3.7"
+source = "git+https://github.com/rust-lang-nursery/log#cb9fa28812ac27c9cadc4e7b18c221b561277289"
+```
+
+Notably `log` from crates.io *is not mentioned at all here*, and crucially so!
+Additionally Cargo has the fully resolved version of the `log` augmentation
+available to it, down to the sha of what to check out.
+
+When Cargo rebuilds from this `Cargo.lock` it will not query the registry for
+versions of `log`, instead seeing that there's an exact dependency on the git
+repository (from the `Cargo.lock`) and the repository is listed as an
+augmentation, so it'll follow that pointer.
+
+## Impact on `[replace]`
+
+The `[augment]` section in the manifest can in many ways be seen as a "replace
+2.0". It is, in fact, strictly more expressive than the current `[replace]`
+section! For example these two sections are equivalent:
+
+```toml
+[replace]
+'log:0.3.7' = { git = 'https://github.com/rust-lang-nursery/log' }
+
+# is the same as...
+
+[augment.crates-io]
+log = { git = 'https://github.com/rust-lang-nursery/log' }
+```
+
+This is not accidental! The intial development of the `[augment]` feature was
+actually focused on prepublishing dependencies and was called `[prepublish]`,
+but while discussing it a conclusion was reached that `[prepublish]` already
+allowed replacing existing versions in a registry, but issued a warning when
+doing so. It turned out that without a warning we ended up having a full-on
+`[replace]` replacement!
```

Comment:
Yeah this sentence makes me think the discussion of warnings on line 189 was supposed to have been removed...